### PR TITLE
Feature/8512 create site service

### LINKED ISSUE #8512: [Site Creation] CreateSite Service
Create service responsible for sending all the data gathered from the user to the server.
- uses the 'sites/new` endpoint as the current site creation flow 
- shows a notification when the app is in the background

### PR DESCRIPTION
**Pull request by malinajirka**
Fixes #8512 

I've broken down the `NewSiteCreationService` into several classes and completely refactored it so it's a bit more testable. Due to number of changes I'd suggest to review it as a new service instead of trying to compare it to the previous version.

`NewSiteCreationServiceManager.kt` - contains most of the business logic. It's job is to move across  and to execute`NewSiteCreationSteps`.

`NewSiteCreationService` - extends already existing and unmodified `AutoForeground` which takes care of displaying notifications when the app is in the background and dismissing them when the app comes to the foreground. Also automatically sends service's state updated through `NewSiteCreationService.updateState(...)` to event bus so all clients Activity/Fragment/VM or whoever wants can update its state accordingly.

Basically the relation between NewSiteCreationService - NewSiteCreationServiceManager is the same as a relation between View - VM.

To test:
It's not tied to the UI yet, so it's up to you if you want to do some initial tests now or leave it for another PR. If you want to test it now, you need to insert something like this `NewSiteCreationService.createSite(this, null, NewSiteCreationServiceData(123, 123, "title", "tagline", "fiuhiq3ryq823ryq3287dya83wyfawr338yt873r"))` somewhere and ideally turn the `dryRun` flag in `CreateSiteUseCase` to `true`.

### CODE REVIEW

File: WordPress/src/main/java/org/wordpress/android/ui/sitecreation/creation/CreateSiteUseCase.kt [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+    suspend fun createSite(siteData: NewSiteCreationServiceData, languageWordPressId: String): OnNewSiteCreated {
+        if (continuation != null) {
+            throw IllegalStateException("Create site request has already been sent.")
+        }
+        return suspendCoroutine { cont ->
+            val newSitePayload = NewSitePayload(
+                    siteData.siteSlug,
+                    siteData.siteTitle ?: "",
+                    languageWordPressId,
+                    PUBLIC,
```
[the remark this answers is not in the document]
Comment (malinajirka) [reply WordPress/src/main/java/org/wordpress/android/ui/sitecreation/creation/CreateSiteUseCase.kt]: Good point - fixed in [2f6aeb2](https://github.com/wordpress-mobile/WordPress-Android/pull/8795/commits/2f6aeb208c137c7f6c7848d1aeebc3cdda5de52a)

File: WordPress/src/main/java/org/wordpress/android/ui/sitecreation/creation/NewSiteCreationServiceManager.kt [hunk 2]
```diff
@@ -0,0 +1,158 @@
...
+
+        isRetry = previousState != null
+
+        val phaseToExecute = if (isRetry) {
+            NewSiteCreationStep.valueOf(previousState!!)
+        } else {
+            IDLE.nextPhase()
+        }
+
+        if (NewSiteCreationServiceState(phaseToExecute, null).isTerminal) {
```
[the remark this answers is not in the document]
Comment (malinajirka) [reply WordPress/src/main/java/org/wordpress/android/ui/sitecreation/creation/NewSiteCreationServiceManager.kt]: Fixed in [3811082](https://github.com/wordpress-mobile/WordPress-Android/pull/8795/commits/381108270b9ebe03f3000cc684838a9643c5c742)

File: WordPress/src/main/java/org/wordpress/android/ui/sitecreation/creation/NewSiteCreationServiceState.kt [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+        return step.name
+    }
+
+    enum class NewSiteCreationStep(val progressPercent: Int = 0) {
+        IDLE,
+        CREATE_SITE(25),
+        SUCCESS,
+        FAILURE;
+
+        fun nextPhase(): NewSiteCreationStep = values()[ordinal + 1]
```
[the remark this answers is not in the document]
Comment (malinajirka) [reply WordPress/src/main/java/org/wordpress/android/ui/sitecreation/creation/NewSiteCreationServiceState.kt]: Yep, you are right. I've just copy pasted it without thinking about it. Fixed in [52705f1](https://github.com/wordpress-mobile/WordPress-Android/pull/8795/commits/52705f1e597cfc4893b16463836e36051ffb7660)

>If we will, I'd convert this to a sealed class so we can have more logic.

I've just removed the method for now as we most probably (and hopefully) won't have any more steps. I haven't converted it to a `sealed class` as [we are putting into a bundle ](https://github.com/wordpress-mobile/WordPress-Android/blob/91ceab88e5a0f7eeff07aeb2734f5ed515a0819f/WordPress/src/main/java/org/wordpress/android/ui/sitecreation/creation/NewSiteCreationService.kt#L112)and retrieving it using `valueOf(..)`. Moreover we don't need any logic for now.

File: WordPress/src/main/java/org/wordpress/android/ui/sitecreation/creation/NewSiteCreationService.kt [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+            intent.putExtra(ARG_DATA, data)
+
+            if (retryFromState != null) {
+                intent.putExtra(ARG_RESUME_PHASE, retryFromState.stepName)
+            }
+
+            context.startService(intent)
+        }
+
+        fun clearSiteCreationServiceState() {
```
[the remark this answers is not in the document]
Comment (malinajirka) [reply WordPress/src/main/java/org/wordpress/android/ui/sitecreation/creation/NewSiteCreationService.kt]: Unfortunately I haven't come up with a better way how to clear the state before the service is started, so I'm calling this method from `onCreate` in the SitePreview fragment - coming in the next PR.

File: WordPress/src/main/java/org/wordpress/android/ui/sitecreation/creation/NewSiteCreationServiceState.kt [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+        return step == NewSiteCreationStep.SUCCESS || isError
+    }
+
+    override fun getStepName(): String {
+        return step.name
+    }
+
+    enum class NewSiteCreationStep(val progressPercent: Int = 0) {
+        IDLE,
+        CREATE_SITE(25),
```
[the remark this answers is not in the document]
Comment (malinajirka) [reply WordPress/src/main/java/org/wordpress/android/ui/sitecreation/creation/NewSiteCreationServiceState.kt]: Constant added in [1601dbd](https://github.com/wordpress-mobile/WordPress-Android/pull/8795/commits/1601dbd443bb980941411762c7e37b1e25cb82e2).

>Better yet, should the percentages be dynamically calculated?

I've discussed this with Sylvester and we've decided having a static value is enough for v1. 

> I am guessing it's supposed to be 100 right now, since it's the only step.

It's the percentage on the progress bar when the Site Creation is in progress. Since we want to indicate something is happening but it's not done yet, we've went for 25%.

@SylvesterWilmott I guess Oguz has a point -> should we completely remove the progress bar for v1 perhaps? As it seems quite useless.

[the remark this answers is not in the document]
Comment (SylvesterWilmott) [reply WordPress/src/main/java/org/wordpress/android/ui/sitecreation/creation/NewSiteCreationServiceState.kt]: I mean, we could switch for an indeterminate progress bar just to show something is happening.

File: WordPress/src/main/java/org/wordpress/android/util/LocaleManager.java [hunk 6]
```diff
@@ -71,18 +74,49 @@ public static boolean isSameLanguage(@NonNull String language) {
...
         SharedPreferences prefs = PreferenceManager.getDefaultSharedPreferences(context);
         return prefs.getString(LANGUAGE_KEY, LanguageUtils.getCurrentDeviceLanguageCode());
     }
 
+    /**
+     * Convert the device language code (codes defined by ISO 639-1) to a Language ID.
+     * Language IDs, used only by WordPress, are integer values that map to a language code.
+     * http://bit.ly/2H7gksN
+     **/
+    public static @NonNull String getLanguageWordPressId(Context context) {
```
[the remark this answers is not in the document]
Comment (malinajirka) [reply WordPress/src/main/java/org/wordpress/android/util/LocaleManager.java]: Yep, I just extracted it to a Util class. 

> I don't have any context

Me neither 😞. I just didn't want to "waste" time on figuring it out.

File: WordPress/src/test/java/org/wordpress/android/ui/sitecreation/creation/NewSiteCreationServiceManagerTest.kt [hunk 7]
```diff
@@ -0,0 +1,217 @@
...
+    @Mock lateinit var tracker: NewSiteCreationTracker
+    @Mock lateinit var dispatcher: Dispatcher
+    @Mock lateinit var serviceListener: NewSiteCreationServiceManagerListener
+
+    private lateinit var manager: NewSiteCreationServiceManager
+
+    @Before
+    fun setUp() {
+        manager = NewSiteCreationServiceManager(useCase, dispatcher, tracker, TEST_DISPATCHER)
+        SUCCESS_EVENT.newSiteRemoteId = NEW_SITE_REMOTE_ID
```
[the remark this answers is not in the document]
Comment (malinajirka) [reply WordPress/src/test/java/org/wordpress/android/ui/sitecreation/creation/NewSiteCreationServiceManagerTest.kt]: Yep I agree - Actually I wanted to ask you if we could change that in FluxC -> add both `dryRun` and `newSiteRemoteId` to the constructor. Wdyt?

File: WordPress/src/main/java/org/wordpress/android/ui/sitecreation/creation/NewSiteCreationServiceNotification.kt [hunk 8]
```diff
@@ -0,0 +1,46 @@
+package org.wordpress.android.ui.sitecreation.creation
+
+import android.app.Notification
+import android.content.Context
+import org.wordpress.android.R
+import org.wordpress.android.util.AutoForegroundNotification
+
+object NewSiteCreationServiceNotification {
+    private const val channelResId = R.string.notification_channel_normal_id
```
[the remark this answers is not in the document]
Comment (malinajirka) [reply WordPress/src/main/java/org/wordpress/android/ui/sitecreation/creation/NewSiteCreationServiceNotification.kt]: I'm not sure I follow. I don't think we should allow the client to show each notification in a different channel.

File: WordPress/src/main/java/org/wordpress/android/ui/sitecreation/creation/NewSiteCreationServiceState.kt (hunk 5 above)
[the remark this answers is not in the document]
Comment (malinajirka) [reply WordPress/src/main/java/org/wordpress/android/ui/sitecreation/creation/NewSiteCreationServiceState.kt]: I've make the progress indeterminate in Removed progress in [1e30994](https://github.com/wordpress-mobile/WordPress-Android/pull/8795/commits/1e309948aa1cd409ae1287c3317b36751fd3ca56)

File: WordPress/src/test/java/org/wordpress/android/ui/sitecreation/creation/NewSiteCreationServiceManagerTest.kt (hunk 7 above)
[the remark this answers is not in the document]
Comment (malinajirka) [reply WordPress/src/test/java/org/wordpress/android/ui/sitecreation/creation/NewSiteCreationServiceManagerTest.kt]: Fixed in [d7ea5ac](https://github.com/wordpress-mobile/WordPress-Android/pull/8795/commits/d7ea5ac92202234e982da2a2eed9f87f9127ec5d)

File: WordPress/src/main/java/org/wordpress/android/ui/sitecreation/creation/NewSiteCreationServiceData.kt [hunk 9]
```diff
@@ -0,0 +1,15 @@
+package org.wordpress.android.ui.sitecreation.creation
+
+import android.annotation.SuppressLint
+import android.os.Parcelable
+import kotlinx.android.parcel.Parcelize
+
+@Parcelize
+@SuppressLint("ParcelCreator")
```
[the remark this answers is not in the document]
Comment (oguzkocer) [reply WordPress/src/main/java/org/wordpress/android/ui/sitecreation/creation/NewSiteCreationServiceData.kt]: Oh, interesting, thanks for the context!

File: WordPress/src/test/java/org/wordpress/android/ui/sitecreation/creation/CreateSiteUseCaseTest.kt [hunk 10]
```diff
@@ -0,0 +1,112 @@
...
+        assertThat(payload.siteName).isEqualTo(DUMMY_SITE_DATA.siteSlug)
+        assertThat(payload.siteTitle).isEqualTo(DUMMY_SITE_DATA.siteTitle)
+        // TODO uncomment when the API is ready
+//        assertThat(payload.segmentId).isEqualTo(DUMMY_SITE_DATA.segmentId)
+//        assertThat(payload.verticalId).isEqualTo(DUMMY_SITE_DATA.verticalId)
+//        assertThat(payload.tagLine).isEqualTo(DUMMY_SITE_DATA.siteTagLine)
+    }
+
+    @Test
+    fun verifyDryRunIsFalse() = test {
```
[the remark this answers is not in the document]
Comment (oguzkocer) [reply WordPress/src/test/java/org/wordpress/android/ui/sitecreation/creation/CreateSiteUseCaseTest.kt]: > It's funny that you brought this up on this specific method. I think I don't usually write these types of methods - tbh I think this is my first in WPAndroid.

This might not have been the best test to bring it up, I probably should have done it in the verticals PR, because even now when I look at the tests for the Verticals VM, I get really overwhelmed and don't know what's going on. I guess I'll have a better understanding of it when I implement the domains VM unit tests and will be interesting choice between following the verticals example and taking a different approach.

> We are giving control over those attributes to the clients even though I don't think they should ever change them. Wdyt?

I don't fully agree with this. These are fully reusable components, a different part of the app could have an entirely different set of requirements. I'd say if it's in the payload, the use case should have it as well, because the whole idea with use cases are to wrap FluxC calls, not hide them. In specific cases, we might want to remove them from the payloads and the use cases together, which I could totally get behind.

> But I guess this is really more about a personal style.

Right :)

File: WordPress/src/test/java/org/wordpress/android/ui/sitecreation/creation/NewSiteCreationServiceManagerTest.kt (hunk 7 above)
[the remark this answers is not in the document]
Comment (oguzkocer) [reply WordPress/src/test/java/org/wordpress/android/ui/sitecreation/creation/NewSiteCreationServiceManagerTest.kt]: > Actually I wanted to ask you if we could change that in FluxC -> add dryRun, newSiteRemoteId and error to the constructor. Wdyt?

Sure thing. Do you mind opening an issue with low priority for this with details of what needs to be changed in WPAndroid so it's easier to find it for me? I don't think we should focus on this right now.

File: WordPress/src/test/java/org/wordpress/android/ui/sitecreation/creation/CreateSiteUseCaseTest.kt (hunk 10 above)
[the remark this answers is not in the document]
Comment (malinajirka) [reply WordPress/src/test/java/org/wordpress/android/ui/sitecreation/creation/CreateSiteUseCaseTest.kt]: ok, you are right, I agree.

> because the whole idea with use cases are to wrap FluxC calls

Offtopic: I don't think we need to limit the use of `use-cases` just to the transformation of FluxC events to coroutines. I think we could theoretically put there any logic. We haven't needed it yet, but we might in the future.

File: WordPress/src/test/java/org/wordpress/android/ui/sitecreation/creation/NewSiteCreationServiceManagerTest.kt (hunk 7 above)
[the remark this answers is not in the document]
Comment (malinajirka) [reply WordPress/src/test/java/org/wordpress/android/ui/sitecreation/creation/NewSiteCreationServiceManagerTest.kt]: Hmm, so I've checked other payloads in FluxC and this is not the only case where the attributes are not in the constructor (none of the payloads in SiteStore have them). It's probably not worth the risk of breaking something. However, thank you for your willingness.

File: WordPress/src/test/java/org/wordpress/android/ui/sitecreation/creation/CreateSiteUseCaseTest.kt (hunk 10 above)
[the remark this answers is not in the document]
Comment (malinajirka) [reply WordPress/src/test/java/org/wordpress/android/ui/sitecreation/creation/CreateSiteUseCaseTest.kt]: >This might not have been the best test to bring it up, I probably should have done it in the verticals PR, because even now when I look at the tests for the Verticals VM, I get really overwhelmed and don't know what's going on. I guess I'll have a better understanding of it when I implement the domains VM unit tests and will be interesting choice between following the verticals example and taking a different approach.

I'm looking forward to see your approach:)!

File: WordPress/src/test/java/org/wordpress/android/ui/sitecreation/creation/NewSiteCreationServiceManagerTest.kt (hunk 7 above)
[the remark this answers is not in the document]
Comment (oguzkocer) [reply WordPress/src/test/java/org/wordpress/android/ui/sitecreation/creation/NewSiteCreationServiceManagerTest.kt]: Right, especially the older ones don't have it in the constructor. It wouldn't break anything, but probably not worth the time spent for us right now. It'd be great if we can fix them in batches.

_So much to do in hack weeks 🙀 _

File: WordPress/src/test/java/org/wordpress/android/ui/sitecreation/creation/CreateSiteUseCaseTest.kt (hunk 10 above)
[the remark this answers is not in the document]
Comment (oguzkocer) [reply WordPress/src/test/java/org/wordpress/android/ui/sitecreation/creation/CreateSiteUseCaseTest.kt]: > Offtopic: I don't think we need to limit the use of use-cases just to the transformation of FluxC events to coroutines. I think we could theoretically put there any logic. We haven't needed it yet, but we might in the future.

Sure, as long as we don't make it our goal to hide issues in FluxC in the use-cases, I am fine with that. The only thing I want is to improve the FluxC side when necessary instead of introducing a layer for it.

[the remark this answers is not in the document]
Comment (malinajirka) [reply WordPress/src/test/java/org/wordpress/android/ui/sitecreation/creation/CreateSiteUseCaseTest.kt]: Yep, that's a good point and I agree!
